### PR TITLE
NNS1-3350 create visibility change modal

### PR DESCRIPTION
**Pull request by coskucinkilic**
# Motivation

Let users change their neurons visibility

# Changes

1. Style changes to render action button correctly in AdvanceSection
2. Add NnsChangeNeuronVisibilityButton 
3. Update NnsNeuronPublicVisibilityAction to use the new button and use isControllable around the button.
4. Update en.json and types
5. Add ChangeNeuronVisibilityModal 
6. Update NnsNeuronModals  to include new modal 
7. Update busy.store to add new initiator type
8. Create ChangeNeuronVisibilityModal.page-object
9. Fix type issues in NeuronContextActionsTest
10. Fix type issues in modal.mock
11. Update screenshots

# Tests
1 update NnsNeuronPublicVisibilityAction.spec
 - handle isControllable neuron display differences

2 Add NnsChangeNeuronVisibilityButton.spec ( needs update)
- should display correct text for public and private neurons
- should open modal on click

3 Add ChangeNeuronVisibilityModal.spec
  -  checks modals presence
  -  making sure that the correct title is displayed based on neuron visibility type
  -  calls changeNeuronVisibility on confirm click
  -  start and stop busy indicator
  -  shows correct success message and close the modal after success both for public and private neurons
  -  should handle error  changing neuron visiblity 
  
4 Screenshots have been updated

# Todos

- [ ] Add entry to changelog (if necessary).

### CODE REVIEW

File: frontend/src/lib/modals/neurons/ChangeNeuronVisibilityModal.svelte [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    {isPublic
+      ? $i18n.neuron_detail.change_neuron_make_neuron_private_confirmation
+      : $i18n.neuron_detail.change_neuron_make_neuron_public_confirmation}
+  </h4>
+
+  <p class="description">
+    {isPublic
+      ? $i18n.neuron_detail.change_neuron_make_neuron_private_description
+      : $i18n.neuron_detail.change_neuron_make_neuron_public_description}
+    <a
```
Comment (dskloetd) [inline frontend/src/lib/modals/neurons/ChangeNeuronVisibilityModal.svelte:69]: If you construct a sentence by concatenating 2 translatable messages, the result isn't really translatable.

I just discussed with Andrew and he's fine with changing the whole "Learn more ... here." sentence to just "Learn more" and make that the link. Then we don't have to dynamically construct a sentence.

(We currently don't translate the app, but if one day we do, it would be good if there aren't too many glitches.)

File: frontend/src/lib/modals/neurons/ChangeNeuronVisibilityModal.svelte [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  import { toastsSuccess } from "$lib/stores/toasts.store";
+
+  export let neuron: NeuronInfo;
+
+  const dispatcher = createEventDispatcher();
+  const close = () => {
+    dispatcher("nnsClose");
+  };
+
+  let applyToAllNeurons = false;
```
Comment (dskloetd) [inline frontend/src/lib/modals/neurons/ChangeNeuronVisibilityModal.svelte:19]: Do I see correctly that this functionality isn't implemented?
It better to do it in a separate PR but then let's do it entirely in a separate PR and not partially in this PR.
So let's remove the Checkbox and variable here.

File: frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+import * as busyServices from "$lib/stores/busy.store";
+import { toastsSuccess } from "$lib/stores/toasts.store";
+import { renderModal } from "$tests/mocks/modal.mock";
+import { mockNeuron } from "$tests/mocks/neurons.mock";
+import { ChangeNeuronVisibilityModalPo } from "$tests/page-objects/ChangeNeuronVisibilityModal.page-object";
+import { JestPageObjectElement } from "$tests/page-objects/jest.page-object";
+import { NeuronVisibility } from "@dfinity/nns";
+import { waitFor } from "@testing-library/svelte";
+
+vi.mock("$lib/services/neurons.services", () => ({
```
Comment (dskloetd) [inline frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts:12]: We prefer only mocking the API layer.
(I'm sorry if you copied this from an old test that still uses older practices.)

File: frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+  toastsSuccess: vi.fn(),
+}));
+
+let spyConsoleError;
+
+describe("ChangeNeuronVisibilityModal", () => {
+  beforeEach(() => {
+    spyConsoleError = vi.spyOn(console, "error");
+  });
+  afterEach(() => {
```
Comment (dskloetd) [inline frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts:26]: Using afterEach is less safe than always using beforeEach.
Can you move the contents of `afterEach` to `beforeEach`?

File: frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+    return {
+      po: ChangeNeuronVisibilityModalPo.under(
+        new JestPageObjectElement(container)
+      ),
+      nnsClose,
+    };
+  };
+
+  beforeEach(() => {
```
Comment (dskloetd) [inline frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts:49]: Is there a reason why you made 2 separate `beforeEach`s?

File: frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+  it("should call changeNeuronVisibility service on confirm click", async () => {
+    const { po } = await renderComponent();
+
+    const confirmButton = po.getConfirmButton();
+
+    await confirmButton.click();
+
+    expect(changeNeuronVisibility).toHaveBeenCalledWith({
+      neurons: [mockNeuron],
+      makePublic: true,
```
Comment (dskloetd) [inline frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts:85]: The expected `makePublic` argument should depend on the neuron that was rendered, right?

File: frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    const confirmButton = po.getConfirmButton();
+
+    await confirmButton.click();
+
+    expect(startBusySpy).toHaveBeenCalledWith({
+      initiator: "change-neuron-visibility",
+      labelKey: "neuron_detail.change_neuron_visibility_loading",
+    });
+
+    await waitFor(() => {
```
Comment (dskloetd) [inline frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts:101]: Using `waitFor` is fragile.
Can you try changing it to
```
await runResolvedPromises();
expect(stopBusySpy).toHaveBeenCalledWith("change-neuron-visibility");
```

File: frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+      visibility: NeuronVisibility.Private,
+    };
+
+    const { po, nnsClose } = await renderComponent(privateNeuron);
+
+    const confirmButton = po.getConfirmButton();
+
+    await confirmButton.click();
+
+    await waitFor(() => {
```
Comment (dskloetd) [inline frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts:118]: Please use `await runResolvedPromises();` instead of `waitFor`.

Also below.

File: frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+import { JestPageObjectElement } from "$tests/page-objects/jest.page-object";
+import { NeuronVisibility } from "@dfinity/nns";
+import { waitFor } from "@testing-library/svelte";
+
+vi.mock("$lib/services/neurons.services", () => ({
+  changeNeuronVisibility: vi.fn().mockResolvedValue({ success: true }),
+}));
+
+vi.mock("$lib/stores/toasts.store", () => ({
+  toastsSuccess: vi.fn(),
```
Comment (dskloetd) [inline frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts:17]: There's no need to mock `toastSuccess`. You can inspect the real toastsStore directly.
If it's easy to use real objects instead of mocking, we prefer not mocking.

Example:
https://github.com/dfinity/nns-dapp/blob/55173ee68739016a787f0c38020dea51a995caa3/frontend/src/tests/lib/pages/NnsProposalDetail.spec.ts#L71

Also make sure to reset the `toastsStore` in `beforeEach`.

File: frontend/src/tests/lib/components/neuron-detail/NeuronContextActionsTest.svelte [hunk 10]
```diff
@@ -6,10 +6,10 @@
   } from "$lib/types/nns-neuron-detail.context";
   import { NNS_NEURON_CONTEXT_KEY } from "$lib/types/nns-neuron-detail.context";
   import type { NeuronInfo } from "@dfinity/nns";
-  import { setContext, SvelteComponent } from "svelte";
+  import { setContext, type ComponentType } from "svelte";
   import { writable } from "svelte/store";
 
-  export let testComponent: typeof SvelteComponent;
+  export let testComponent: ComponentType;
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/neuron-detail/NeuronContextActionsTest.svelte:12]: Why this change?

File: frontend/src/tests/lib/components/neuron-detail/NnsNeuronPublicVisibilityAction.spec.ts [hunk 11]
```diff
@@ -10,14 +16,28 @@ describe("NnsNeuronPublicVisibilityAction", () => {
...
+    ...mockNeuron,
+    fullNeuron: {
+      ...mockNeuron.fullNeuron,
+      controller: mockIdentity.getPrincipal().toText(),
+    },
+  };
+
+  beforeEach(() => {
+    resetAccountsForTesting();
+    vi.spyOn(authStore, "subscribe").mockImplementation(mockAuthStoreSubscribe);
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/neuron-detail/NnsNeuronPublicVisibilityAction.spec.ts:35]: Please don't mock stores (I'm sorry we still have existing tests that do this.)
Use `resetIdentity` if you want to test as if the user is signed in or `setNoIdentity` as if no user is signed in.

File: frontend/src/tests/lib/components/neuron-detail/actions/NnsChangeNeuronVisibilityButton.spec.ts [hunk 12]
```diff
@@ -0,0 +1,52 @@
+import NnsChangeNeuronVisibilityButton from "$lib/components/neuron-detail/actions/NnsChangeNeuronVisibilityButton.svelte";
+import { mockNeuron } from "$tests/mocks/neurons.mock";
+import { ButtonPo } from "$tests/page-objects/Button.page-object";
+import { JestPageObjectElement } from "$tests/page-objects/jest.page-object";
+import { NeuronVisibility } from "@dfinity/nns";
+import { render } from "@testing-library/svelte";
+import NeuronContextActionsTest from "../NeuronContextActionsTest.svelte";
+
+describe("NnsChangeNeuronVisibilityButton", () => {
+  afterEach(() => {
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/neuron-detail/actions/NnsChangeNeuronVisibilityButton.spec.ts:10]: Please always use `beforeEach` instead of `afterEach`.

See https://docs.google.com/document/d/1sVXb7IXToniKdHsEnTb3KFz1BQ8IkrVNV59bEiJWIUM/edit#heading=h.lre0iwedyyu

File: frontend/src/tests/lib/components/neuron-detail/actions/NnsChangeNeuronVisibilityButton.spec.ts [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        neuron: { ...mockNeuron, visibility },
+        testComponent: NnsChangeNeuronVisibilityButton,
+      },
+    });
+    return {
+      po: ButtonPo.under({
+        element: new JestPageObjectElement(container),
+        testId: "change-neuron-visibility-button",
+      }),
+      queryByTestId,
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/neuron-detail/actions/NnsChangeNeuronVisibilityButton.spec.ts:26]: How about returning a second page object (for the modal) so we test everything through page objects?

File: frontend/src/tests/mocks/modal.mock.ts [hunk 14]
```diff
@@ -27,14 +27,13 @@ export const waitModalIntroEnd = async ({
...
   component,
   props,
 }: {
-  component: typeof SvelteComponent;
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  props?: Record<string, any>;
-}): Promise<RenderResult<SvelteComponent>> => {
+  component: ComponentType<T>;
+  props?: ComponentProps<T>;
+}): Promise<RenderResult<T>> => {
```
Comment (dskloetd) [inline frontend/src/tests/mocks/modal.mock.ts:36]: What's the reason for these changes?

File: frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts (hunk 3 above)
Comment (coskucinkilic) [reply frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts]: Updated.

File: frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts (hunk 9 above)
Comment (coskucinkilic) [reply frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts]: updated

File: frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts (hunk 4 above)
Comment (coskucinkilic) [reply frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts]: updated.

File: frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts (hunk 5 above)
Comment (coskucinkilic) [reply frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts]: i saw it as well it has been fixed now.

File: frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts (hunk 6 above)
Comment (coskucinkilic) [reply frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts:109]: yes this is based on the neuron rendered. i explicitly pass private neuron now to make it more obvious. do you want me to also write a test for makePublic false version?

File: frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts (hunk 7 above)
Comment (coskucinkilic) [reply frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts]: Deleted all waitFors and used await runResolvedPromises();

File: frontend/src/tests/mocks/modal.mock.ts (hunk 14 above)
Comment (coskucinkilic) [reply frontend/src/tests/mocks/modal.mock.ts:36]: The renderModal was expecting Record<sting,any props> which is quite weak, right now the props type depends on the component passed to it, so we can use the actual props of the component we will render.

File: frontend/src/lib/modals/neurons/ChangeNeuronVisibilityModal.svelte (hunk 1 above)
Comment (coskucinkilic) [reply frontend/src/lib/modals/neurons/ChangeNeuronVisibilityModal.svelte:68]: I updated it please review again

File: frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts (hunk 6 above)
Comment (dskloetd) [reply frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts:109]: Sure, let's do it as well.

File: frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts [hunk 15]
```diff
@@ -46,48 +79,41 @@ describe("ChangeNeuronVisibilityModal", () => {
...
-      neurons: [mockNeuron],
+    expect(changeNeuronVisibilitySpy).toHaveBeenCalledWith({
+      neurons: [privateNeuron],
       makePublic: true,
     });
   });
 
   it("should start and stop busy indicator when changing visibility", async () => {
-    const { po } = await renderComponent();
+    queryNeuronSpy.mockImplementation(async () => privateNeuron);
```
Comment (dskloetd) [inline frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts:114]: This can be simplified to `queryNeuronSpy.mockResolvedValue(privateNeuron);`

below as well

File: frontend/src/lib/modals/neurons/ChangeNeuronVisibilityModal.svelte (hunk 1 above)
Comment (dskloetd) [reply frontend/src/lib/modals/neurons/ChangeNeuronVisibilityModal.svelte:68]: Yes, thanks!

File: frontend/src/lib/components/neuron-detail/NnsNeuronAdvancedSection.svelte [hunk 16]
```diff
@@ -144,6 +144,7 @@
   }
 
   .visibility-action-container {
+    width: 100%;
```
[the remark this answers is not in the document]
Comment (dskloetd) [reply frontend/src/lib/components/neuron-detail/NnsNeuronAdvancedSection.svelte:147]: For the future, would be nice to have such independent fixies in a separate PR so it's clear what it's for.
Adding small fixes into large PRs can make the large PR more confusing and hide why things are the way they are.

File: frontend/src/tests/lib/components/neuron-detail/NeuronContextActionsTest.svelte (hunk 10 above)
Comment (dskloetd) [reply frontend/src/tests/lib/components/neuron-detail/NeuronContextActionsTest.svelte:12]: How come this started causing problems with your PR and not earlier?
Or was it always giving errors but not in a way that caused CI to fail?
Where are you seeing these error? In your IDE? Or when running some command?

File: frontend/src/tests/mocks/modal.mock.ts (hunk 14 above)
Comment (dskloetd) [reply frontend/src/tests/mocks/modal.mock.ts:36]: This this start giving problems with your changes or is it unrelated to the PR?

File: frontend/src/tests/lib/components/neuron-detail/NeuronContextActionsTest.svelte (hunk 10 above)
Comment (coskucinkilic) [reply frontend/src/tests/lib/components/neuron-detail/NeuronContextActionsTest.svelte:12]: These errors was always there in my IDE, i think it wasn't causing CI to fail, i have checked if it is only for my changes, but all uses of both functions that i have updated were giving error all the time. 

There was more but i didn't change them as they are not related to this PR. If you like i can look into type issues and fix them on a separate PR.

File: frontend/src/tests/mocks/modal.mock.ts (hunk 14 above)
Comment (coskucinkilic) [reply frontend/src/tests/mocks/modal.mock.ts:36]: it was giving issue for all uses in the repo in my IDE. nothing specific to my pr.

File: frontend/src/tests/lib/components/neuron-detail/NeuronContextActionsTest.svelte (hunk 10 above)
Comment (dskloetd) [reply frontend/src/tests/lib/components/neuron-detail/NeuronContextActionsTest.svelte:12]: If this is not related to or necessary for this PR, can you remove these changes from this PR?
Then feel free to make a separate PR where the issue is clearly explained in the PR description.
Ideally it would also add a lint check which shows the problem and makes sure it doesn't happen again.

File: frontend/src/tests/mocks/modal.mock.ts (hunk 14 above)
Comment (dskloetd) [reply frontend/src/tests/mocks/modal.mock.ts:36]: Then please move this to a separate PR as well so it can have proper attention.

File: frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts (hunk 6 above)
Comment (coskucinkilic) [reply frontend/src/tests/lib/modals/neurons/ChangeNeuronVisibilityModal.spec.ts:109]: added

File: frontend/src/tests/lib/components/neuron-detail/NeuronContextActionsTest.svelte (hunk 10 above)
Comment (coskucinkilic) [reply frontend/src/tests/lib/components/neuron-detail/NeuronContextActionsTest.svelte:12]: deleted

File: frontend/src/tests/mocks/modal.mock.ts (hunk 14 above)
Comment (coskucinkilic) [reply frontend/src/tests/mocks/modal.mock.ts:36]: reverted

**Review by dskloetd — APPROVED** [review]
